cmd/key-helper/importKey.go

```diff
@@ -140,7 +145,89 @@ func wrapPrivateKey(wrappingKey string, privateKeyPath string) (string, error) {
 	return base64Ciphertext, nil
 }
 
-func importKeyToTransit(ctx context.Context, client *vault.Client, ciphertext string, keyName string) error {
+// Checks PrivateKey type and bitlength in EC, PKCS1 and PKCS8 containers
+func checkKeyType(keyPath string) (string, string, error) {
+	// Key types supported by Notary project specification
+	supportedRSAKeys := map[int]string{
+		2048: "rsa-2048",
+		3072: "rsa-3072",
+		4096: "rsa-4096",
+	}
+	supportedECDSAKeys := map[string]string{
+		"P-256": "ecdsa-p256",
+		"P-384": "ecdsa-p384",
+		"P-521": "ecdsa-p521",
+	}
+	digestAlgorithm := map[string]string{
+		"ecdsa-p256": "SHA256",
+		"ecdsa-p384": "SHA384",
+		"ecdsa-p521": "SHA512",
+		"rsa-2048":   "SHA256",
+		"rsa-3072":   "SHA384",
+		"rsa-4096":   "SHA512",
+	}
+
+	keyString, err := os.ReadFile(keyPath)
+	if err != nil {
+		return "", "", err
+	}
+	pemDecoded, _ := pem.Decode([]byte(keyString))
+
+	switch pemDecoded.Type {
+	case "EC PRIVATE KEY":
```

Comment:
I would not, in general, suggest this. PEM blobs can have a variety of markers and usually it is better to opportunistically read various blocks than it is to rely on the typed header.

I guess this is a difference in philosophies between OpenBao and Notary though:

https://github.com/notaryproject/notation-core-go/blob/93218d92c134a02dfa325c074374cdfb4d108fab/x509/key.go#L34-L49

vs

https://github.com/openbao/openbao/blob/6736b400f6a5efdbf7247462d3b21fa49432f634/sdk/helper/certutil/helpers.go#L244-L251 

into

https://github.com/openbao/openbao/blob/6736b400f6a5efdbf7247462d3b21fa49432f634/sdk/helper/certutil/helpers.go#L210-L242 

which ignores the PEM header entirely, but gives its expected canonical header.

@FeynmanZhou do you know who to ask about the project philosophies around parsing keys here?

Reply:
By reusing the already parsed version in `wrapPrivateKey`, you can avoid all the below parsing-of-DER code, but obviously will leave the length determination code. My 2c., but this should be a stronger design and less prone to breaking (one source to update to add new key parsing, rather than once here and once in the shared library). Obviously new `key_type` correlation will need to be handled manually here, but we can default to err out.